### PR TITLE
add Expires header + allow the use of regex to extract value from XCallID body

### PR DESCRIPTION
**Pull request by games130**
Added to extract Expires header from SIP message

XCallID to be able to use regex to further extract correct valid. Example for P-Charging-Vector

P-Charging-Vector: icid-value="somevalue";orig-ioi=host.domain.com

will want to extract "somevalue" instead of the whole string.
Suggested format to add the regex. Regex need to have group construct ( ):
current toml: AlegIDs = ["X-CID","P-Charging-Vector"]
suggested toml: AlegIDs = ["X-CID","P-Charging-Vector,icid-value=\"(.*)\""]

> Could you please rebase it so I can evaluate only the change you need. Currently I don't have that much time to assist here. Another thing that I noticed is that you spilt the Header and compile the regex everytime. Doing it this way is quite slow and slould be improved.

@negbie I have created another PR to fix the messy PR previously so now it only show one commit. Also i have updated the code to try not to compile the regex everytime. do let me know your comments.

### CODE REVIEW

File: sipparser/parser.go [hunk 1]
```diff
@@ -246,11 +250,23 @@ func (s *SipMsg) addHdr(str string) {
...
 		case s.hdr == "Expires":
+			s.Expires = s.hdrv
 		default:
 			if len(s.XHeader) > 0 {
 				for i := range s.XHeader {
 					if s.hdr == s.XHeader[i] {
-						s.XCallID = s.hdrv
+						var filter *regexp.Regexp
+						var ok bool
+						filter, ok = config.CompileStore.RegexMap[s.XHeader[i]]
```
Comment (negbie) [inline sipparser/parser.go]: You can just use
filter, ok := config.CompileStore.RegexMap[s.XHeader[i]]

and drop the vars above

File: sipparser/parser.go [hunk 2]
```diff
@@ -246,11 +250,23 @@ func (s *SipMsg) addHdr(str string) {
...
+						if ok {
+							if filter.MatchString(s.hdrv){
+								s.XCallID = filter.FindStringSubmatch(s.hdrv)[1]
+							} else {
+								s.XCallID = "ERR: "+s.hdrv
```
Comment (negbie) [inline sipparser/parser.go]: Do we really need this?
s.XCallID = "ERR: "+s.hdrv

File: cmd/heplify-server/heplify-server.go [hunk 3]
```diff
@@ -44,6 +45,22 @@ func init() {
 	} else {
 		fmt.Println("Could not find toml config file, use flag defaults.", err)
 	}
+	
+	
+	if len(config.Setting.AlegIDs) > 0 {
```
Comment (negbie) [inline cmd/heplify-server/heplify-server.go]: Could you put this block into a function or so you can reuse it in the test.

File: sipparser/parser_test.go [hunk 4]
```diff
@@ -122,6 +125,39 @@ func TestParseMsg(t *testing.T) {
...
+		"Contact: something\r\nAuthorization: Digest username=\"foobaruser124\", realm=\"FOOBAR\"\r\nProxy-Authorization: Digest username=\"foobaruser124\", realm=\"FOOBAR\"\r\n" +
+		"X-RTP-Stat: CS=0;PS=1433;ES=1525;OS=229280;SP=0/0;SO=0;QS=-;PR=1522;ER=1525;OR=243520;CR=0;SR=0;QR=-;PL=0,0;BL=0;LS=0;RB=0/0;SB=-/-;EN=PCMA,FAX;DE=PCMA;JI=23,2;DL=20,20,21;IP=83.138.49.179:7082,102.183.157.163:25132\r\nX-RTP-Stat-Add: DQ=31;DSS=0;DS=0;PLCS=288;JS=1\r\n\r\nv=0\r\no=Dialogic_SDP 1452654 0 IN IP4 0.0.0.0\r\ns=Dialogic-SIP\r\nc=IN IP4 4.71.122.135\r\nt=0 0\r\nm=audio 11676 RTP/AVP 0 101\r\na=rtpmap:0 PCMU/8000\r\na=rtpmap:101 telephone-event/8000\r\na=fmtp:101 0-15\r\na=silenceSupp:off - - - -\r\na=ptime:20\r\n"
+	
+	if len(x) > 0 {
```
Comment (negbie) [inline sipparser/parser_test.go]: See me comment above. Just put this into a function.

File: sipparser/parser.go (hunk 1 above)
Comment (games130) [reply sipparser/parser.go]: done

File: sipparser/parser.go (hunk 2 above)
Comment (games130) [reply sipparser/parser.go]: its optional, so i removed it

File: cmd/heplify-server/heplify-server.go (hunk 3 above)
Comment (games130) [reply cmd/heplify-server/heplify-server.go]: done

File: sipparser/parser_test.go (hunk 4 above)
Comment (games130) [reply sipparser/parser_test.go]: done

**Comment by games130** [comment]
i have updated the code, based on your comments

**Comment by negbie** [comment]
@games130 
Thank you very much!

**Pull request merged**